packages/validation/src/rule-provider.ts

```diff
@@ -628,11 +636,17 @@ export class ValidationMessageProvider implements IValidationMessageProvider {
   }
 
   public getMessage(rule: IValidationRule): Interpolation | PrimitiveLiteralExpression {
-    const parsedMessage = this.registeredMessages.get(rule);
-    if (parsedMessage !== void 0) { return parsedMessage; }
+    const messageKey = rule.messageKey;
+    const lookup = this.registeredMessages.get(rule);
+    if (lookup != null) {
+      const parsedMessage = lookup.get(explicitMessageKey) ?? lookup.get(messageKey);
+      if (parsedMessage !== void 0) { return parsedMessage; }
+    }
+
+    // In case of stateful rule, the message key is the message itself.
+    if (rule._isStateful) return this.setMessage(rule, messageKey);
```

Comment:
the rule provider knowing too much about the internal of the rule is quite undesirable. do we need to resort to this? It seems we already take dependency on `StateRule` inside this RuleProvider, maybe just be with it here?
```suggestion
    if (rule instanceof StateRule) return this.setMessage(rule, rule.message);
```